### ISSUE TITLE
Connection unusable after SELECTs interrupted by OS signal

### ISSUE THREAD

**Issue by mactyr** [psycopg/psycopg2#471]
# Description #

I am working on a long-running python process that performs a lot of database access (mostly reads, occasional writes). Sometimes it may be necessary to terminate the process before it finishes (e.g. by using the `kill` command) and when this happens I would like to log a value to the database indicating that the particular run was canceled. (I am also logging the occurrence to a log file; I would like to have the information in both places.)

I have found that if I interrupt the process while the database connection is active, the connection becomes unusable; specifically, it hangs the process if I try to use it in any way.

# Minimum working example #

This snippet reproduces the problem reliably.

The table `test` in the database has two columns, `id` (serial) and `message` (text). I prepopulated it with one row so the `UPDATE` statement below would have something to change.

```python
import psycopg2
import sys
import signal

    
pg_host = 'localhost'
pg_user = 'redacted'
pg_password = 'redacted'
pg_database = 'test_db'


def write_message(msg):
    print "Writing: " + msg
    cur.execute("UPDATE test SET message = %s WHERE id = 1", (msg,))
    conn.commit()


def signal_handler(signal, frame):
    write_message('Interrupt!')
    sys.exit(0)
signal.signal(signal.SIGINT, signal_handler)
signal.signal(signal.SIGTERM, signal_handler)


if __name__ == '__main__':
    conn = psycopg2.connect(host=pg_host, user=pg_user, password=pg_password, database=pg_database)
    cur = conn.cursor()

    write_message("Starting")
    for i in xrange(10000):
        # I press ^C somewhere in here
        cur.execute("SELECT * FROM test")
        cur.fetchall()
    write_message("Finishing")
```

When I run this script without interruption, it completes as expected. That is, the row in the database is updated to say "Starting" then "Finishing".

If I press `ctrl-C` during the loop indicated by the comment, python hangs indefinitely. It no longer responds to keyboard input, and the process has to be killed from elsewhere. Looking in my postgresql log, the `UPDATE` statement with "Interrupted!" is never received by the database server.

If I add a debugging breakpoint at the beginning of signal_handler() I can see that doing almost anything with the database connection at that point causes the same hang. Trying to `execute` a `SELECT`, issuing a `conn.rollback()`, `conn.commit()`, `conn.close()` or `conn.reset()` all cause the hang. Executing `conn.cancel()` does not cause a hang, but it doesn't improve the situation; subsequent use of the connection still causes a hang. If I remove the database access from `write_message()` then the script is able to exit gracefully when interrupted, so the hang is definitely database connection related.

Also worth noting: if I change the script so that I am interrupting something other than database activity, it works as desired, logging "Interrupted!" to the database. E.g., if I replace the `for i in xrange(10000)` loop with a simple `sleep(10)` and interrupt that, it works fine. So the problem seems to be specifically related to interrupting psycopg2 with a signal while it is performing database access, *then* trying to use the connection.

# Workaround #

I've discovered that if I create an entirely new connection with `psycopg2.connect()` after the interrupt and am careful not to access the old one, I can update the database using the new connection. That's what I'm doing for now, but it feels untidy -- it seems like there should be a way to salvage the existing connection, or at least close it down cleanly. (Also, ideally I would get an exception when I try to use the connection in this bad state rather than a hang.)

# Environment #

 - OS X 10.11.6
 - python 2.7.11
 - psycopg2 2.6.1
 - postgresql 9.5.1.0

For what it's worth, this was originally [posted](http://stackoverflow.com/questions/39215527/psycopg2-connection-unusable-after-selects-interrupted-by-os-signal) on StackOverflow but hasn't received any responses after being up for a few days.

**Comment by dvarrazzo** [comment]
What you are trying to do depends largely on the behaviour of the underlying OS. I don't think you can reliable do network communication in a signal handler. At most you can use the handler to set a variable or put something in a queue, then call `write_message()` from the "main loop".

Try using a connection in green mode, e.g. use `psycopg2.extensions.set_wait_callback(psycopg2.extras.wait_select)`: this would the OS `select()` wrapped by Python's `select` module instead of the one the libpq uses: there is a chance that it would react better to the signals. In any case the connection would remain in an unusable state and you will need a new connection (or maybe `reset()`) to use it again.

**Comment by mactyr** [comment]
Ok, thanks for the clarification.

In my environment, using `psycopg2.extensions.set_wait_callback(psycopg2.extras.wait_select)` does improve the situation; when I interrupt the for loop I wind up with an exception from the `cur.execute()` line in `write_message()` ("ProgrammingError: execute cannot be used while an asynchronous query is underway") rather than a hang. Interestingly, though, if I try to do a `conn.reset()` at the beginning of `signal_handler()` while using `wait_select` I get a hang again! So, when using `wait_select`, `execute()` gives me an exception but `reset()` still gives me a hang, which aligns with your advice that it's best just to not touch the old connection once the signal handler has started.

As an aside, I noticed in [the documentation](http://initd.org/psycopg/docs/faq.html#faq-interrupt-query) that there had been some change to `wait_select` in psycopg2 2.6.2 whereas I had been using 2.6.1. I upgraded to 2.6.2 and in my case all the behavior I tested is exactly the same as in 2.6.1.

I'll close this issue since we've documented the available mitigation / workarounds and there doesn't seem to be anything more to do within psycopg2. Thanks again!
